pkg/storage/storagewrappers/boundedconcurrency.go

```diff
@@ -0,0 +1,82 @@
+package storagewrappers
+
+import (
+	"context"
+	"time"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const timeWaitingSpanAttribute = "time_waiting"
+
+var _ storage.RelationshipTupleReader = (*boundedConcurrencyTupleReader)(nil)
+
+var (
+	timeWaitingHistogram = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name:    "datastore_bounded_read_delay_ms",
+		Help:    "Time spent waiting for Read, ReadUserTuple and ReadUsersetTuples calls to the datastore",
+		Buckets: []float64{10, 25, 50, 100, 1000, 5000}, // milliseconds. Upper bound is config.UpstreamTimeout
```

Comment:
Please configure this as a native histogram so we have better resolution than these predefined buckets, and also we'll want to tighten up the low range of these buckets since hopefully, on avg, the amount of waiting that we'll have will be in the single-digit milliseconds. So I recommend at least having a `1ms`, `3ms`, and `5ms` in the range.

See:
https://promcon.io/2022-munich/talks/native-histograms-in-prometheus/
https://github.com/auth0/sandcastle/blob/main/internal/storage/dynamodb/utils/retryableclient.go#L37-L39